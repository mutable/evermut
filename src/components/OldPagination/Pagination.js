import React from 'react';
import PropTypes from 'prop-types';
import { Pane, Select, option, Button } from 'evergreen-ui';
import Loader from '../Loader';

const PAGE_LIMIT = 10;
const DEFAULT_PAGE = 1;

class OldPagination extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      pageIndex: props.currentPage || DEFAULT_PAGE,
      pageLimit: props.pageLimit || PAGE_LIMIT
    };
  }

  handleChange = (event) => {
    const index = parseInt(event.target.value, 10);

    this.clicked(index);
  };

  _getPages = (countOfPages) => {
    const pages = [];
    if (countOfPages === 0) {
      pages[0] = 0;
    } else {
      for (let i = 1; i <= countOfPages; i++) {
        pages.push(i);
      }
    }

    return pages.map(page => (
      <option key={page}>
        { page }
      </option>
    ));
  }

  clicked(index) {
    const { pageLimit } = this.state;
    if (this.props.onClick) this.props.onClick(index, pageLimit);
    this.setState({ pageIndex: index });
  }

  render() {
    const { pageIndex, pageLimit } = this.state;
    const { count, classes, translate, loading } = this.props;

    const countOfPages = Math.ceil(count / pageLimit);
    const pages = this._getPages(countOfPages);

    const backActive = (pageIndex <= 1);
    const nextActive = (pageIndex >= pages.length);
    const nextPage = pageIndex + 1;
    const prevPage = pageIndex - 1;

    return ( loading ? <Loader /> : 
      <Pane>
        <Button
          disabled={backActive}
          onClick={() => this.clicked(prevPage)}
        >
          Back
        </Button>
        <Select
          value={pageIndex}
          onChange={e => this.handleChange(e)}
          disabled={countOfPages < 2}
        >
          {pages}
        </Select>
        <Button
          disabled={nextActive}
          onClick={() => this.clicked(nextPage)}
        >
          Next
        </Button>
      </Pane>
    );
  }
}

OldPagination.defaultProps = {
  pageLimit: 10,
  currentPage: 1,
  loading: false
}

OldPagination.propTypes = {
  loading: PropTypes.bool,
  count: PropTypes.number.isRequired,
  pageLimit: PropTypes.number,
  currentPage: PropTypes.number,
  onClick: PropTypes.func.isRequired,
};

export default OldPagination;
