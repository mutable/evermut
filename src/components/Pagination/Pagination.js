import React from 'react';
import PropTypes from 'prop-types';
import { Pane, Button, IconButton, Text } from 'evergreen-ui';
import Loader from '../Loader';

const DISTANCE = 10;
const MARGIN = 15;
const HEIGHT = 25;
const COLOR = '#425a70';
const FONT = 11;

class Pagination extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      pageIndex: props.pageIndex,
      pageLimit: props.limit,
      countOfPages: Math.ceil(props.count / props.limit),
      count: props.count,
      pages: [],
      front: '',
      back: '...',
    };
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.count !== nextProps.count) {
      return { countOfPages: Math.ceil(nextProps.count / prevState.pageLimit), count: nextProps.count };
    }
    return null;
  }

  clicked(index) {
    const { pageLimit, pageIndex } = this.state;

    if (this.props.onClick) this.props.onClick(index, pageLimit);
    this._getEllipses(index);
    this.setState({ pageIndex: index });
  }

  clickedDouble(isNext) {
  	const { countOfPages, pageLimit } = this.state;
  	const pageIndex = isNext ? countOfPages : 1;
    if (this.props.onClick) this.props.onClick(pageIndex, pageLimit);
  	this.setState({ pageIndex });
  }

  _getEllipses(index) {
  	const { countOfPages } = this.state;
  	let front = '...'
  	let back = '...'
  	if(index >= 1 && index <= countOfPages) {
  		if((index - 1) <= 2) {
  			front = ''
  		}
  		if(countOfPages - (index + 1) <= 1) {
  			back = ''
  		}
  	}
  	this.setState({ front, back });
  }

  _getPages(index) {
  	const { countOfPages, pageIndex, count } = this.state;
  	const pages = [];
    
    if(count === 0) {
        pages.push(1);
    } else if (countOfPages <= 4) {
  		for (let i = 1; i <= countOfPages; i++) {
  			pages.push(i);
  		}
  	} else {
		 	if ((index + 1) > countOfPages) {
	  		pages.push(index - 2, index - 1, index);
	  	} else if ((index - 1) < 1) {
	  		pages.push(index, index + 1, index + 2);
	  	} else {
	  		pages.push(index - 1, index, index + 1);
	  	}
  	}
		return pages;
  }

  pageButton(number) {
    const  { countOfPages, pageIndex } = this.state;
    return <Button
      disabled={countOfPages < 2}
      fontWeight={pageIndex === number ? "bold" : 'normal'}
      background={pageIndex === number ? "#edf0f2" : 'white'}
      appearance="minimal"
      onClick={() => this.clicked(number)}
      paddingLeft={DISTANCE}
      paddingRight={DISTANCE}
      marginLeft={DISTANCE/2}
      marginRight={DISTANCE/2}
      height={HEIGHT}
      fontSize={FONT}
      color={COLOR}
    >
      {number}
    </Button>
  }

	isIncluded(array, n){
		const { pages, countOfPages } = this.state;

		const item = array.findIndex(i => (i === n))

		if (item !== -1 ) return false;
		return true;
	}

  render() {
    const { pageIndex, pageLimit, countOfPages, pages, front, back } = this.state;
    const { count, classes, translate, loading } = this.props;

    const backActive = (pageIndex <= 1);
    const nextActive = (pageIndex >= countOfPages);
    const nextPage = pageIndex + 1;
    const prevPage = pageIndex - 1;

    const _pages = this._getPages(pageIndex);

    return ( loading ? <Loader /> : 
      <Pane display='inline-flex' marginRight={MARGIN}>
        <IconButton appearance="minimal" height={HEIGHT} disabled={backActive} onClick={() => this.clicked(prevPage)} icon="chevron-left" />
        <Pane>
        	{countOfPages <= 4 && _pages.map(page => this.pageButton(page))}
        	{countOfPages > 4 &&
        		<Pane>
	        		{this.isIncluded(_pages, 1) && this.pageButton(1)}
		        	<Text fontSize={FONT} color={COLOR}>{front}</Text>
		        	 {
		        		_pages.map(page => this.pageButton(page))
							 }
		        	<Text fontSize={FONT} color={COLOR}>{back}</Text>
					    {this.isIncluded(_pages, countOfPages) && this.pageButton(countOfPages)}
		        </Pane>
	        }
        </Pane>
        <IconButton appearance="minimal" height={HEIGHT} fontSize={FONT} disabled={nextActive} onClick={() => this.clicked(nextPage)} icon="chevron-right" />
      </Pane>
    );
  }
}

Pagination.defaultProps = {
  limit: 10,
  pageIndex: 1,
  loading: false
}

Pagination.propTypes = {
  count: PropTypes.number.isRequired,
  limit: PropTypes.number,
  pageIndex: PropTypes.number,
  onClick: PropTypes.func.isRequired
}

export default Pagination;