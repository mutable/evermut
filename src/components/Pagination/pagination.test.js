import React from 'react';
import renderer from 'react-test-renderer';

import Pagination from '../Pagination';

describe('<Pagination />', () => {
  it('component with min requirements render correctly', () => {
    const component = renderer.create(<Pagination
      loading={false}
      count={30}
      onClick={jest.fn()}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('component advanced render correctly', () => {
    const component = renderer.create(
      <Pagination
        loading={false}
        count={300}
        limit={25}
        pageIndex={3}
        onClick={jest.fn()}
      />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});