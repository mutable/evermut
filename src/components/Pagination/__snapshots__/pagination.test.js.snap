// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> component advanced render correctly 1`] = `
<div
  className="css-nil ðŸ“¦dspl_inline-flex ðŸ“¦mr_15px ðŸ“¦box-szg_border-box"
>
  <button
    className="css-1ii3p2c ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_0px ðŸ“¦pl_0px ðŸ“¦ml_0px ðŸ“¦mr_0px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦w_25px ðŸ“¦just-cnt_center ðŸ“¦box-szg_border-box"
    onClick={[Function]}
  >
    <svg
      className="ðŸ“¦w_12px ðŸ“¦h_12px ðŸ“¦box-szg_border-box"
      data-icon="chevron-left"
      style={
        Object {
          "fill": "#66788A",
        }
      }
      viewBox="0 0 16 16"
    >
      <path
        d="M7.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71l-4 4C5.11 7.47 5 7.72 5 8c0 .28.11.53.29.71l4 4a1.003 1.003 0 001.42-1.42L7.41 8z"
        fillRule="evenodd"
      />
    </svg>
  </button>
  <div
    className="css-nil ðŸ“¦box-szg_border-box"
  >
    <div
      className="css-nil ðŸ“¦box-szg_border-box"
    >
      <button
        className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_normal ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_11ev722 ðŸ“¦box-szg_border-box"
        onClick={[Function]}
      >
        1
      </button>
      <span
        className="ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦box-szg_border-box"
      >
        
      </span>
      <button
        className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_normal ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_11ev722 ðŸ“¦box-szg_border-box"
        onClick={[Function]}
      >
        2
      </button>
      <button
        className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_bold ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_1ofpdvj ðŸ“¦box-szg_border-box"
        onClick={[Function]}
      >
        3
      </button>
      <button
        className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_normal ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_11ev722 ðŸ“¦box-szg_border-box"
        onClick={[Function]}
      >
        4
      </button>
      <span
        className="ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦box-szg_border-box"
      >
        ...
      </span>
      <button
        className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_normal ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_11ev722 ðŸ“¦box-szg_border-box"
        onClick={[Function]}
      >
        12
      </button>
    </div>
  </div>
  <button
    className="css-1ii3p2c ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_0px ðŸ“¦pl_0px ðŸ“¦ml_0px ðŸ“¦mr_0px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦w_25px ðŸ“¦just-cnt_center ðŸ“¦box-szg_border-box"
    onClick={[Function]}
  >
    <svg
      className="ðŸ“¦w_12px ðŸ“¦h_12px ðŸ“¦box-szg_border-box"
      data-icon="chevron-right"
      style={
        Object {
          "fill": "#66788A",
        }
      }
      viewBox="0 0 16 16"
    >
      <path
        d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
        fillRule="evenodd"
      />
    </svg>
  </button>
</div>
`;

exports[`<Pagination /> component with min requirements render correctly 1`] = `
<div
  className="css-nil ðŸ“¦dspl_inline-flex ðŸ“¦mr_15px ðŸ“¦box-szg_border-box"
>
  <button
    className="css-1ii3p2c ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_0px ðŸ“¦pl_0px ðŸ“¦ml_0px ðŸ“¦mr_0px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦w_25px ðŸ“¦just-cnt_center ðŸ“¦box-szg_border-box"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="ðŸ“¦w_12px ðŸ“¦h_12px ðŸ“¦box-szg_border-box"
      data-icon="chevron-left"
      style={
        Object {
          "fill": "#66788A",
        }
      }
      viewBox="0 0 16 16"
    >
      <path
        d="M7.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71l-4 4C5.11 7.47 5 7.72 5 8c0 .28.11.53.29.71l4 4a1.003 1.003 0 001.42-1.42L7.41 8z"
        fillRule="evenodd"
      />
    </svg>
  </button>
  <div
    className="css-nil ðŸ“¦box-szg_border-box"
  >
    <button
      className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_bold ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_1ofpdvj ðŸ“¦box-szg_border-box"
      onClick={[Function]}
    >
      1
    </button>
    <button
      className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_normal ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_11ev722 ðŸ“¦box-szg_border-box"
      onClick={[Function]}
    >
      2
    </button>
    <button
      className="css-1ii3p2c ðŸ“¦color_425a70 ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_normal ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_10px ðŸ“¦pl_10px ðŸ“¦ml_5px ðŸ“¦mr_5px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦bg_11ev722 ðŸ“¦box-szg_border-box"
      onClick={[Function]}
    >
      3
    </button>
  </div>
  <button
    className="css-1ii3p2c ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_11px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_25px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_0px ðŸ“¦pl_0px ðŸ“¦ml_0px ðŸ“¦mr_0px ðŸ“¦mb_0px ðŸ“¦h_25px ðŸ“¦pst_relative ðŸ“¦dspl_flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦w_25px ðŸ“¦just-cnt_center ðŸ“¦box-szg_border-box"
    onClick={[Function]}
  >
    <svg
      className="ðŸ“¦w_12px ðŸ“¦h_12px ðŸ“¦box-szg_border-box"
      data-icon="chevron-right"
      style={
        Object {
          "fill": "#66788A",
        }
      }
      viewBox="0 0 16 16"
    >
      <path
        d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
        fillRule="evenodd"
      />
    </svg>
  </button>
</div>
`;
